python/dgl/data/utils.py

```diff
@@ -482,3 +484,198 @@ def add_nodepred_split(dataset, ratio, ntype=None):
+
+    Returns
+    ----------
+    split_masks : dict
+        A python dict storing the mask names as keys and the corresponding
+        node mask arrays as values.
+    """
+
+    num_nodes = len(property_values)
+    part_sizes = np.round(num_nodes * np.array(part_ratios)).astype(int)
+    part_sizes[-1] -= np.sum(part_sizes) - num_nodes
+
+    generator = np.random.RandomState(random_seed)
+    permutation = generator.permutation(num_nodes)
+
+    node_indices = np.arange(num_nodes)[permutation]
+    property_values = property_values[permutation]
+    in_distribution_size = np.sum(part_sizes[:3])
+
+    node_indices_ordered = node_indices[np.argsort(property_values)]
+    node_indices_ordered[: in_distribution_size] = generator.permutation(
+        node_indices_ordered[: in_distribution_size]
+    )
+
+    sections = np.cumsum(part_sizes)
+    node_split = np.split(node_indices_ordered, sections)[:-1]
+    mask_names = [
+        "in_train_mask",
+        "in_valid_mask",
+        "in_test_mask",
+        "out_valid_mask",
+        "out_test_mask",
+    ]
+    split_masks = {}
+
+    for mask_name, node_indices in zip(mask_names, node_split):
+        split_mask = idx2mask(node_indices, num_nodes)
+        split_masks[mask_name] = generate_mask_tensor(split_mask)
+
+    return split_masks
+
+
+def add_node_property_split(
+    dataset, part_ratios, property_name, ascending=True, random_seed=None
+):
+    """Create a data split with a distributional shift based on some node property.
+
+    It splits each graph in the given dataset into training, ID and OOD validation, ID and OOD
+    testing parts for transductive node prediction task with structural distributional shifts.
+    As a result, it creates 5 associated node mask arrays for each graph:
+        - 3 for the ID nodes: ``'in_train_mask'``, ``'in_valid_mask'``, ``'in_test_mask'``,
+        - and 2 for the OOD nodes: ``'out_valid_mask'``, ``'out_test_mask'``.
+
+    Following `Evaluating Robustness and Uncertainty of Graph Models Under Structural
+    Distributional Shifts <https://arxiv.org/abs/2302.13875v1>`__, this function implements
+    3 particular strategies for inducing distributional shifts in graph — based on
+    **popularity**, **locality** or **density**.
+
+    Parameters
+    ----------
+    dataset : DGLDataset or list of :class:`DGLGraph`
+        The dataset to induce structural distributional shift in.
+    part_ratios : list
+        A list of 5 ratio values for training, ID validation, ID test,
+        OOD validation and OOD test parts. The values must sum to 1.0.
+    property_name : str
+        The node property name to split the dataset by. Must be
+        ``'popularity'``, ``'locality'`` or ``'density'``.
+    ascending : bool, optional
+        Whether to sort nodes in the ascending order of a particular
+        node property, so that more shifted OOD nodes have greater values
+        of the computed property (default: True)
```

Comment:
"so that more shifted OOD nodes have greater values of the computed property (default: True)" -> "so that nodes with greater values of the property are considered to be OOD"

Reply:
Fixed.